### PR TITLE
SEQNG-431 Don't configure GMOS mechanisms if they are already in place.

### PR DESCRIPTION
**Pull request by jluhrs**

### CODE REVIEW

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 1]
```diff
@@ -56,6 +56,12 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
 
   implicit val disperserOrderEncoder: EncodeEpicsValue[DisperserOrder, String] = EncodeEpicsValue(_.sequenceValue)
 
+  implicit val disperserOrderEncoderInt: EncodeEpicsValue[DisperserOrder, Int] = EncodeEpicsValue{
+    case Order.ZERO => 0
+    case Order.ONE  => 1
+    case Order.TWO  => 2
+  }
```
Comment (jluhrs) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: I could have used the order, but I don't think that is a good idea.

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 2]
```diff
@@ -109,39 +115,50 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
     _ <- DC.setCcdYBinning(encode(dc.bi.y))
   } yield ()
 
-
+  private def smartSetParam[A: Equal](v: A, get: => Option[A], set: SeqAction[Unit]): SeqAction[Unit] =
+    get.flatMap(g => (g =/= v).option(set)).getOrElse(SeqAction.void)
```
Comment (jluhrs) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: I first defined `set` as `A => SeqAction[Unit]`, and passed `v` to it. The problem is that, in some cases, the type of the parameter is not always the type of the associated status.

Comment (cquiroz) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: Looks nice

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 3]
```diff
@@ -109,39 +115,50 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
...
 
   def setFilters(f: T#Filter): SeqAction[Unit] = {
     val (filter1, filter2) = encoders.filter.encode(f)
 
-    CC.setFilter1(filter1) *> CC.setFilter2(filter2)
+    smartSetParam(filter1, GmosEpics.instance.filter1, CC.setFilter1(filter1)) *>
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: Will the `*>` make this run in parallel?

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 4]
```diff
@@ -109,39 +115,50 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
...
 
   def setFPU(cc: GmosFPU): SeqAction[Unit] = {
+    def inBeamDecode(v: Int): String = if (v===0) InBeamVal else OutOfBeamVal
+
     def builtInFPU(fpu: T#FPU): SeqAction[Unit] = {
       val (fpuName, beam) = encoders.fpu.encode(fpu)
 
-      fpuName.fold(SeqAction.void)(CC.setFpu) *>
-        beam.fold(SeqAction.void)(CC.setInBeam)
+      fpuName.fold(SeqAction.void)(v => smartSetParam(v, GmosEpics.instance.fpu, CC.setFpu(v))) *>
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: It seems this call is duplicated on line 160

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 5]
```diff
@@ -154,14 +171,19 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
...
-  } yield ()
+  def setCCConfig(cc: GmosController.Config[T]#CCConfig): SeqAction[Unit] = {
+    val stage = encoders.stageMode.encode(cc.stage)
+
+    for {
+      _ <- setFilters(cc.filter)
+      _ <- setDisperser(cc.disperser)
+      _ <- setFPU(cc.fpu)
+      _ <- smartSetParam(stage, GmosEpics.instance.stageMode, CC.setStageMode(stage))
+      _ <- CC.setDtaXOffset(cc.dtaX.intValue.toDouble*PixelsToMicrons)
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: I bet you meant to have a space at each side of `*`

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 6]
```diff
@@ -154,14 +171,19 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
...
+      _ <- cc.useElectronicOffset.fold(CC.setElectronicOffsetting(0))(e =>
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: Perhaps a clearer constant would be better than the `0` in `setElectronicOffseting`

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala [hunk 7]
```diff
@@ -109,39 +115,50 @@ class GmosControllerEpics[T<:GmosController.SiteDependentTypes](encoders: GmosCo
...
-    CC.setDisperser(encoders.disperser.encode(d.disperser)) *>
-      CC.setDisperserMode(disperserMode) *>
-      d.order.filter(_ => d.disperser != Disperser.MIRROR).fold(SeqAction.void)(o => CC.setDisperserOrder(encode(o))) *>
-      d.lambda.filter(_ => d.disperser != Disperser.MIRROR && !d.order.contains(Order.ZERO)).fold(SeqAction.void)(o => CC.setDisperserLambda(encode(o)))
+    val disperserMode0 = "WLEN"
+    val disperserMode1 = "SEL"
+    val disperser = encoders.disperser.encode(d.disperser)
+    def disperserModeDecode(v : Int): String = if(v===0) disperserMode0 else disperserMode1
+
+    smartSetParam(disperser, GmosEpics.instance.disperser, CC.setDisperser(disperser)) *>
```
Comment (cquiroz) [inline modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: This is a bit hard to follow, could it be perhaps split into several functions. Up to you

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala (hunk 3 above)
Comment (jluhrs) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: No. It runs the left one, discards the result, and then runs the right one.

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala (hunk 4 above)
Comment (jluhrs) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: You are right. I'll factor it out.

File: modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala (hunk 3 above)
Comment (cquiroz) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: If the first were to fail, would we know?

Comment (jluhrs) [reply modules/edu.gemini.seqexec.server/src/main/scala/edu/gemini/seqexec/server/gmos/GmosControllerEpics.scala]: Yes. Those are `EitherT[Task, SeqexecFailure, Unit]`. The `EitherT` shortcircuits on the left.
